pkg/relay/relay.go

```diff
@@ -0,0 +1,121 @@
+// limitations under the License.
+
+package relay
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+)
+
+type Relay struct {
+	addr          *net.UDPAddr
+	bufferChannel chan string
+	logger        log.Logger
+	packetLength  uint
+}
+
+// NewRelay creates a statsd UDP relay. It can be used to send copies of statsd raw
+// lines to a separate service.
+func NewRelay(l log.Logger, target string, packetLength uint) (*Relay, error) {
+	addr, err := net.ResolveUDPAddr("udp", target)
+	if err != nil {
+		return nil, fmt.Errorf("unable to resolve target %s, err: %w", target, err)
+	}
+
+	c := make(chan string, packetLength)
+
+	r := Relay{
+		addr:          addr,
+		bufferChannel: c,
+		logger:        l,
+		packetLength:  packetLength,
+	}
+
+	// Startup the UDP sender.
+	go r.relayOutput()
+
+	return &r, nil
+}
+
+// relayOutput buffers statsd lines and sends them to the relay target.
+func (r *Relay) relayOutput() {
+	var buffer string
+	var err error
+	relayTimeout := 1 * time.Second
+
+	timeout := time.After(relayTimeout)
+
+	for {
+		select {
+		case <-timeout:
+			level.Debug(r.logger).Log("msg", "Relay buffer timeout, sending packet", "packet", buffer)
+			err = r.sendPacket(buffer)
+			if err != nil {
+				level.Error(r.logger).Log("msg", "Error sending UDP packet", "error", err)
+				return
+			}
+			// Clear out the buffer.
+			buffer = ""
+			timeout = time.After(relayTimeout)
+		case b := <-r.bufferChannel:
+			if uint(len(b)+1+len(buffer)) > r.packetLength {
+				level.Debug(r.logger).Log("msg", "Buffer full, sending packet", "packet", buffer)
+				err = r.sendPacket(buffer)
+				if err != nil {
+					level.Error(r.logger).Log("msg", "Error sending UDP packet", "error", err)
+					return
+				}
+				// Seed the new buffer with the new line.
+				buffer = b + "\n"
+			} else {
+				level.Debug(r.logger).Log("msg", "Adding line to buffer", "line", b)
+				buffer = buffer + b + "\n"
+			}
+		}
+	}
+}
+
+// sendPacket sends a single relay line to the destination target.
+func (r *Relay) sendPacket(buf string) error {
+	if len(buf) == 0 {
+		level.Debug(r.logger).Log("msg", "Empty buffer, nothing to send")
+		return nil
+	}
+	conn, err := net.ListenUDP("udp", nil)
+	if err != nil {
+		return err
+	}
+	level.Debug(r.logger).Log("msg", "Sending packet", "length", len(buf))
+	_, err = conn.WriteToUDP([]byte(buf), r.addr)
+	return err
+}
+
+// RelayLine processes a single statsd line and forwards it to the relay target.
+func (r *Relay) RelayLine(l string) {
+	lineLength := uint(len(l))
+	if lineLength == 0 {
+		level.Debug(r.logger).Log("msg", "Empty line, not relaying")
+		return
+	}
+	if lineLength > r.packetLength-1 {
+		level.Warn(r.logger).Log("msg", "Line longer than max allowed length, dropping line", "length", lineLength, "max", r.packetLength)
+		return
+	}
+	level.Debug(r.logger).Log("msg", "Relaying line", "line", l)
+	r.bufferChannel <- l
```

Comment:
I am worried that this can backpressure and impact ingestion if volume is very high. It's _probably_ fine, but we could make this a [non-blocking write](https://gobyexample.com/non-blocking-channel-operations) and increment a counter when it fails. Would that be worth it?

Reply:
Yea, let me think about this one.

Reply:
How about we try it this way, and I will test how bad it is.

Reply:
works for me!